copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopyBookSource.java

```diff
@@ -0,0 +1,283 @@
+
+  private final CopybookSourceConfig config;
+  private Schema outputSchema;
+
+  public CopybookSource(CopybookSourceConfig copybookConfig) {
+    this.config = copybookConfig;
+  }
+
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(config.parseSchema());
+
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    if (config.schema != null) {
+      outputSchema = config.parseSchema();
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws IOException {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    conf.set(CopybookConstants.COPYBOOK_INPUTFORMAT_CBL_CONTENTS, config.copybookContents);
+    conf.setInt(CopybookConstants.COPYBOOK_INPUTFORMAT_FILE_STRUCTURE, config.fileStructure);
+    conf.set(CopybookConstants.COPYBOOK_INPUTFORMAT_DATA_HDFS_PATH, config.binaryFilePath);
+
+    // Set the input file path for the job
+    CopybookInputFormat.setInputPaths(job, config.binaryFilePath);
+
+    if (config.maxSplitSize != null) {
+      conf.setLong(CopybookConstants.MAX_SPLIT_SIZE_DESCRIPTION, config.maxSplitSize);
+      CopybookInputFormat.setMaxInputSplitSize(job, conf.getLong(CopybookConstants.MAX_SPLIT_SIZE_DESCRIPTION,
+                                                                 CopybookConstants.DEFAULT_MAX_SPLIT_SIZE));
+    }
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(CopybookInputFormat.class, conf)));
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Map<String, AbstractFieldValue>> input,
+                        Emitter<StructuredRecord> emitter)
+    throws Exception {
+    Map<String, AbstractFieldValue> value = input.getValue();
+    if (outputSchema == null) {
+      outputSchema = getSchema(value.keySet());
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    for (Schema.Field field : outputSchema.getFields()) {
+      String fieldName = field.getName();
+      if (value.containsKey(fieldName)) {
+        builder.set(fieldName, parseValue(field, value.get(fieldName)));
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Convert each field to data type as specified by user in the output schema
+   *
+   * @param field Schema.Field object defining the field name and properties
+   * @param value value of field to be converted to datatype specified by user
+   * @return value casted to the required data type
+   * @throws IOException
+   */
+  private Object parseValue(Schema.Field field, AbstractFieldValue value) throws IOException {
+    Schema fieldSchema = field.getSchema();
+    Schema.Type fieldType = fieldSchema.isNullable() ? fieldSchema.getNonNullable().getType() : fieldSchema
```

Comment:
you need to handle the case where value is null.  The way it is now, you will get a null pointer exception